File: Intermediate/3.Merge two sorted arrays into a single sorted array/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

void mergeSortedArrays(const vector<int> &arr1, const vector<int> &arr2, vector<int> &merged) {
    int i = 0, j = 0;

    // Merge both arrays
    while (i < arr1.size() && j < arr2.size()) {
        if (arr1[i] <= arr2[j]) {
            merged.push_back(arr1[i++]);
        } else {
            merged.push_back(arr2[j++]);
        }
    }

    // Copy remaining elements of arr1
    while (i < arr1.size()) {
        merged.push_back(arr1[i++]);
    }

    // Copy remaining elements of arr2
    while (j < arr2.size()) {
        merged.push_back(arr2[j++]);
    }
}

int main() {
    int n1, n2;

    cout << "Enter the number of elements in the first sorted array: ";
    cin >> n1;
    vector<int> arr1(n1);
    cout << "Enter the elements of the first sorted array:\n";
    for (int i = 0; i < n1; i++) {
        cin >> arr1[i];
    }

    cout << "Enter the number of elements in the second sorted array: ";
    cin >> n2;
    vector<int> arr2(n2);
    cout << "Enter the elements of the second sorted array:\n";
    for (int i = 0; i < n2; i++) {
        cin >> arr2[i];
    }

    vector<int> merged;
    mergeSortedArrays(arr1, arr2, merged);

    cout << "Merged sorted array: ";
    for (int num : merged) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: Intermediate/3.Merge two sorted arrays into a single sorted array/Solution.java
import java.util.Scanner;

public class MergeSortedArrays {

    static int[] mergeSortedArrays(int[] arr1, int[] arr2) {
        int n1 = arr1.length, n2 = arr2.length;
        int[] merged = new int[n1 + n2];
        int i = 0, j = 0, k = 0;

        // Merge both arrays
        while (i < n1 && j < n2) {
            if (arr1[i] <= arr2[j]) {
                merged[k++] = arr1[i++];
            } else {
                merged[k++] = arr2[j++];
            }
        }

        // Copy remaining elements of arr1
        while (i < n1) {
            merged[k++] = arr1[i++];
        }

        // Copy remaining elements of arr2
        while (j < n2) {
            merged[k++] = arr2[j++];
        }

        return merged;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of elements in the first sorted array: ");
        int n1 = scanner.nextInt();
        int[] arr1 = new int[n1];
        System.out.println("Enter the elements of the first sorted array:");
        for (int i = 0; i < n1; i++) {
            arr1[i] = scanner.nextInt();
        }

        System.out.print("Enter the number of elements in the second sorted array: ");
        int n2 = scanner.nextInt();
        int[] arr2 = new int[n2];
        System.out.println("Enter the elements of the second sorted array:");
        for (int i = 0; i < n2; i++) {
            arr2[i] = scanner.nextInt();
        }

        int[] merged = mergeSortedArrays(arr1, arr2);

        System.out.print("Merged sorted array: ");
        for (int num : merged) {
            System.out.print(num + " ");
        }
        System.out.println();

        scanner.close();
    }
}
